update: kotlin-test as a single dependency (#2137)
JetBrains/kotlin-web-site: docs/topics/gradle.md

### Set dependencies on test libraries

> The [`kotlin.test` API](https://kotlinlang.org/api/latest/kotlin.test/) is available for testing Kotlin projects on different Kotlin platforms.

Reviewer: "different Kotlin platforms" doesn't tell exactly what platforms you mean. Please write precisely. "All supported platforms" I suppose.

Author: Supported by whom? 
"all platforms supported by Kotlin" I would say.

Reviewer: "**Kotlin** projects on all supported platforms" implies that platforms are supported by Kotlin. No need for more words here.

Author: Indeed) Thank you!

Add the dependency `kotlin-test` to the `commonTest` source set, and the Gradle plugin will infer the corresponding 

> test dependencies for each source set:

Reviewer: Do we mean test source sets here? If we do, then the current wording is not completely true.
If we really mean _each_ source set in the project, then the behavior is strange =)

Author: Just to recheck:
@ilya-g, does Gradle infer the corresponding test dependencies for each **test** source set?

Reviewer 2: It depends on where `kotlin("test")` dependency is added. If in `commonTest`, then all test source sets actualizing `commonTest` also get the respective variant of `kotlin-test` dependency.

### Set dependencies on test libraries

> The [`kotlin.test` API](https://kotlinlang.org/api/latest/kotlin.test/) is available for testing Kotlin projects on 

Author: See how this link looks: https://branch-kotlin-test-as-the-single-dependency.kotlin-web-site.labs.jb.gg/docs/gradle.html#set-dependencies-on-test-libraries

Should I edit the formatting somehow?

Reviewer: Yes, let's move `API` out of the link.

The [`kotlin.test` API](https://kotlinlang.org/api/latest/kotlin.test/) is available for testing different Kotlin projects. 

Add the dependency `kotlin-test` to the `commonTest` source set, and the Gradle plugin will infer the corresponding 
test dependencies for each source set:
* `kotlin-test-common` and `kotlin-test-annotations-common` for common source sets

> * `kotlin-test-junit`/`kotlin-test-junit-5`/`kotlin-test-testng` for JVM source sets

Reviewer: Please avoid using `/` as a conjunction. It's often unclear what you're replacing with it: and, or, something else?
I'd mention only `kotlin-test-junit` in the first sentence because it is actually inferred with `kotlin-test`. Perhaps we can add a second sentence about JUnit 5  and TestNG.

Author: Should I also do sth with ```/``` in [what's new](https://github.com/JetBrains/kotlin-web-site/pull/2192/files#r607146909)?

> * `kotlin-test-js` for Kotlin/JS source sets

Reviewer 2: I don't think that details of dependency names for each platform are useful to readers given that now all should work with the single dependency.

Author: But may it worth to be mentioned for the common understanding? Let's ask @p7nov 's opinion

Reviewer: This may be useful for readers who add these dependencies manually now. They'll learn what exactly is replaced with `kotlin("test")`. But I agree that it's not necessary.

Author: Not necessary but may be useful – then I'll leave it here if nobody is against.

Reviewer 2: ok

Kotlin/Native targets do not require additional test dependencies, and the `kotlin.test` API implementations are built-in.

Additionaly, you can use the `kotlin-test` dependency in any shared or platform-specific source set.

> For Kotlin/JVM Gradle uses JUnit 4 by default; therefore, the `kotlin('test')` dependency resolves to

Reviewer: I'd move this sentence right before the one about JUnit5 and TestNG. Otherwise, the info about JVM frameworks is broken into two parts

Author: Do you mean it is split by the note about the shortcut?
>You can use shorthand for a dependency on a Kotlin module, for example, kotlin('test') 

Or do you also mean the code below which demonstrates this sentence?

I'll deploy to a staging and put a link to it here; let's look at this page on the staging and decide whether these sentences are not alongside.

> the variant for JUnit 4, namely `kotlin-test-junit`:

Reviewer: `the variant for` – is this part necessary? the sentence seems OK if we just drop these words.
Additionally, `variant` is a term that was never introduced in our docs.

Author: Should I fix it in [what's new](https://github.com/JetBrains/kotlin-web-site/pull/2192/files#r607194270) too?:)

    sourceSets {
        commonTest {
            dependencies {

>                 implementation kotlin('test') // This brings all the platforms' dependencies automatically

Reviewer: ```suggestion
                implementation kotlin('test') // This brings all the platform dependencies automatically
```

</tabs>

> > You can use shorthand for a dependency on a Kotlin module, for example, kotlin('test') for "org.jetbrains.kotlin:kotlin-test".

Reviewer: This shorthand has limitations if I remember correctly. If it does, let's write exactly when it's available. Please check

Reviewer: This seems to require an effort and out of this feature's scope. Let's do this separately later: https://youtrack.jetbrains.com/issue/KT-45974

>
{type="note"}

Additionally, you can use the `kotlin-test` dependency in any shared or platform-specific source set.

> For Kotlin/JVM Gradle uses JUnit 4 by default; therefore, the `kotlin('test')` dependency resolves to JUnit 4.

Reviewer: _kotlin("test")_ (double quotes) is better because it works in both Groovy and Kotlin scripts.

Reviewer 2: > the `kotlin('test')` dependency resolves to JUnit 4.

This new wording looks confusing to me. `kotlin("test")` doesn't resolve to JUnit 4, it resolves to the artifact of `kotlin-test` for JUnit 4. We called it "variant" before as it is called in the standard Gradle terminology of variant-aware dependency resolution.

Author: Thank you! I'm going to return back "the `kotlin("test")` dependency resolves to the variant for 
JUnit 4, namely `kotlin-test-junit`"

You can choose JUnit 5 or TestNG by calling 
[`useJUnitPlatform()`]( https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useJUnitPlatform) 

> or [`useTestNG()`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useTestNG).

Reviewer: by calling them where?

Author: Fixed.

> The following example is for a JVM project:

Reviewer: Perhaps we should add an MPP example since we started this section from MPP.
```
kotlin {
    jvm {
        testRuns["test"].executionTask.configure {
            useJUnitPlatform()
        }
    }
}
```

Author: As I remember, on the call we've discussed that there are many server-side users to whom we target the code too. You told it will be cool if there is a ready piece of code that includes as adding the dependency, as calling usage of some platform. Did I understand you incorrectly or did you change your mind? Or maybe you offer to add the additional example next to the existing one?

Reviewer: Yes, you're right.
I suggest adding **one more example** because calling these tasks in MPP is not so trivial as in JVM. The copy-pastable piece of config can save a lot of time for our readers :)

Author: Added the example.

<tabs>

```groovy
dependencies {
    testImplementation 'org.jetbrains.kotlin:kotlin-test'
}

test {
    useTestNG()
}
```

```kotlin
dependencies {
    testImplementation(kotlin("test"))
}

tasks {
    test {

>         useJUnitPlatform()

Reviewer: Different frameworks in Groovy and Kotlin tabs. Please choose one.

Author: Thank you. Just for the history: as we discussed in DM, I would have noticed this mistake by my own, but the review has been made on the draft PR without a call to do it:)

    }
}
```

</tabs>

> See [the tutorial on how to test code using JUnit in JVM](#jvm-test-using-junit.md).

Reviewer: ```suggestion
[Learn how to test code using JUnit on the JVM](jvm-test-using-junit.md).
```
"Learn how to ..." is our common wording for doc page links.

> If the standard variants described above don't cover you needs, you can disable the automatic selection

Reviewer: In general, I'd start this sentence from a more exact case description such as: "If you need to use a different JVM test framework, you can ... and add this framework as a dependency ..."

Author: I fixed but I'm afraid that the new sentence is too long and complex.

>
{type="note"}

You can use the `kotlin-test` dependency in any shared or platform-specific source set as well.

For Kotlin/JVM Gradle uses JUnit 4 by default; therefore, the `kotlin("test")` dependency resolves to the variant for 
JUnit 4, namely `kotlin-test-junit`.

You can choose JUnit 5 or TestNG by calling 
[`useJUnitPlatform()`]( https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useJUnitPlatform) 
or [`useTestNG()`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useTestNG) in the 
test task of your build script.
The following example is for a JVM project:

<tabs>

```groovy
dependencies {
    testImplementation 'org.jetbrains.kotlin:kotlin-test'
}

test {
    useTestNG()
}
```

```kotlin
dependencies {
    testImplementation(kotlin("test"))
}

tasks {
    test {
        useTestNG()
    }
}
```

</tabs>

[Learn how to test code using JUnit on the JVM](jvm-test-using-junit.md).

> If you need to use a different JVM test framework, you can disable the automatic selection of a testing framework by 
> adding the line `kotlin.test.infer.jvm.variant=false` to the project’s `gradle.properties` and add this framework as a 
> dependency to your build script.

Reviewer: This sentence is indeed too long. Let's think about splitting it.
Just an idea:
```suggestion
If you need to use a different JVM test framework, disable the automatic selection of a testing framework by 
adding the line `kotlin.test.infer.jvm.variant=false` to the project’s `gradle.properties`. Once you do this, add this this framework as a Gradle dependency.
```

</tabs>

> You can use shorthand for a dependency on a Kotlin module, for example, kotlin('test') for "org.jetbrains.kotlin:kotlin-test".
>
{type="note"}

You can choose JUnit 5 or TestNG by calling 
[`useJUnitPlatform()`]( https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useJUnitPlatform) 
or [`useTestNG()`](https://docs.gradle.org/current/javadoc/org/gradle/api/tasks/testing/Test.html#useTestNG) in the test task:

 <tabs>

```groovy
tasks {
    test {
        // enable TestNG support
        useTestNG()
        // or
        // enable JUnit Platform (a.k.a. JUnit 5) support
        useJUnitPlatform()
    }
}
```

> ```kotlin

Reviewer: If Groovy and Kotlin snippets are the same, you can leave only one. We usually leave Groovy/

tasks.test {
    // enable TestNG support
    useTestNG()
    // or
    // enable JUnit Platform (a.k.a. JUnit 5) support
    useJUnitPlatform()
}
```

 </tabs>

> You can disable the automatic selection of a testing framework by adding the line `kotlin.test.infer.jvm.variant=false` 

Reviewer 2: It's not a testing framework that is selected automatically, but a variant of `kotlin-test` library.

> You can disable the automatic selection of a variant of `kotlin-test` library by adding the line `kotlin.test.infer.jvm.variant=false` 

Reviewer: And what to do is you disable it?

Author: Supplemented the sentence, but maybe it became too long now.
